src/components/tl/ucp/alltoall/alltoall_onesided.c

```diff
@@ -11,55 +11,362 @@
+        .team = team->params.team,
+        .args.coll_type = UCC_COLL_TYPE_BARRIER,
+    };
+    size_t                       perc_bw =
+        UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.alltoall_onesided_percent_bw;
+    ucc_on_off_auto_value_t      is_get  = 
+        UCC_TL_UCP_TEAM_LIB(tl_team)->cfg.alltoall_onesided_enable_get;
+    ucc_coll_task_t             *barrier_task;
+    ucc_coll_task_t             *a2a_task;
+    ucc_tl_ucp_task_t           *task;
+    ucc_status_t                 status;
+    ucc_tl_ucp_schedule_t       *tl_schedule;
+    size_t                       nelems;
+    size_t                       rate;
+    size_t                       ratio;
+    ucp_ep_h                     ep;
+    ucp_ep_evaluate_perf_param_t param;
+    ucp_ep_evaluate_perf_attr_t  attr;
+    int64_t                      npolls;
+    ucc_sbgp_t                  *sbgp;
 
-    if (ucc_tl_ucp_test_onesided(task, gsize) == UCC_INPROGRESS) {
-        return;
+    ALLTOALL_TASK_CHECK(coll_args->args, tl_team);
+    if (coll_args->args.mask & UCC_COLL_ARGS_FIELD_FLAGS) {
+        if (!(coll_args->args.flags & UCC_COLL_ARGS_FLAG_MEM_MAPPED_BUFFERS)) {
+            tl_error(UCC_TL_TEAM_LIB(tl_team),
+                     "non memory mapped buffers are not supported");
+            status = UCC_ERR_NOT_SUPPORTED;
+            goto out;
+        }
+    }
+    if (!(coll_args->args.mask & UCC_COLL_ARGS_FIELD_MEM_MAP_SRC_MEMH)) {
+        coll_args->args.src_memh.global_memh = NULL;
+    }
+    if (!(coll_args->args.mask & UCC_COLL_ARGS_FIELD_MEM_MAP_DST_MEMH)) {
+        coll_args->args.dst_memh.global_memh = NULL;
+    }
+    status = ucc_tl_ucp_get_schedule(tl_team, coll_args,
+                                     (ucc_tl_ucp_schedule_t **)&tl_schedule);
+    if (ucc_unlikely(UCC_OK != status)) {
+        return status;
+    }
+    schedule = &tl_schedule->super.super;
+    ucc_schedule_init(schedule, coll_args, team);
+    schedule->super.post     = ucc_tl_ucp_alltoall_onesided_sched_start;
+    schedule->super.progress = NULL;
+    schedule->super.finalize = ucc_tl_ucp_alltoall_onesided_sched_finalize;
+
+    sbgp                 = ucc_topo_get_sbgp(tl_team->topo, UCC_SBGP_NODE);
+    task                 = ucc_tl_ucp_init_task(coll_args, team);
+    task->super.finalize = ucc_tl_ucp_alltoall_onesided_finalize;
+    a2a_task             = &task->super;
+
+    status = ucc_tl_ucp_coll_init(&barrier_coll_args, team, &barrier_task);
+    if (status != UCC_OK) {
+        return status;
+    }
+    if (perc_bw > 100) {
+        perc_bw = 100;
+    } else if (perc_bw == 0) {
+        perc_bw = 1;
     }
 
-    pSync[0]           = 0;
-    task->super.status = UCC_OK;
+    nelems             = TASK_ARGS(task).src.info.count;
+    nelems             = nelems / UCC_TL_TEAM_SIZE(tl_team);
+    param.field_mask   = UCP_EP_PERF_PARAM_FIELD_MESSAGE_SIZE;
+    attr.field_mask    = UCP_EP_PERF_ATTR_FIELD_ESTIMATED_TIME;
+    param.message_size = (1 << 20);
```

Comment:
why 1mb specifcally?

Reply:
I think it is better to use the actual send size (token size) since bw is not linear and UCX might use different protocols depending on messages size